swev-id: django__django-14539 fix urlize entity trailing punctuation

casey-brooks opened the pull request:
## Summary
- fix #426 by trimming trailing punctuation based on the unescaped token so entities don't duplicate their suffix
- add regression coverage for escaped/numeric entities and template filter nofollow rendering

## Reproduction & failure
1. utils urlize path:
   ```
   PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_html.TestUtilsHtml.test_urlize --parallel=1
   ```
   ```
   AssertionError: 'Search for <a href=\"http://google.com/?q=1%3C\">google.com/?q=1&lt</a>lt! and see.' != 'Search for <a href=\"http://google.com/?q=1%3C\">google.com/?q=1&lt</a>! and see.'
   ```
2. template filter path:
   ```
   PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_punctuation --parallel=1
   ```
   ```
   AssertionError: 'Search for <a href=\"http://google.com/?q=1%3C\" rel=\"nofollow\">google.com/?q=1&amp;lt</a>lt! and see.' != 'Search for <a href=\"http://google.com/?q=1%3C\" rel=\"nofollow\">google.com/?q=1&lt</a>! and see.'
   ```

## Fix
- compute the trailing punctuation length from the unescaped string and move exactly that many characters from the escaped token into the trail

## Testing
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_html.TestUtilsHtml.test_urlize --parallel=1
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_punctuation template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_parentheses --parallel=1
- flake8 django/utils/html.py tests/utils_tests/test_html.py tests/template_tests/filter_tests/test_urlize.py

## Links
- #426
- django/django#14539

Python 3.12.12 via nix profile; dependencies installed into a venv.

casey-brooks committed:
fix(utils.html): correct entity trailing trim

Refs: #426

Reproduction:
PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_html.TestUtilsHtml.test_urlize --parallel=1

Failure excerpt:
AssertionError: 'Search for <a href="http://google.com/?q=1%3C">google.com/?q=1&lt</a>lt! and see.' != 'Search for <a href="http://google.com/?q=1%3C">google.com/?q=1&lt</a>! and see.'

Template filter reproduction:
PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_punctuation --parallel=1

Failure excerpt:
AssertionError: 'Search for <a href="http://google.com/?q=1%3C" rel="nofollow">google.com/?q=1&amp;lt</a>lt! and see.' != 'Search for <a href="http://google.com/?q=1%3C" rel="nofollow">google.com/?q=1&lt</a>! and see.'

casey-brooks commented:
Test & Lint Summary:
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_html.TestUtilsHtml.test_urlize --parallel=1 (1 test, 1 passed)
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_punctuation template_tests.filter_tests.test_urlize.UrlizeTests.test_urlize_entity_trailing_parentheses --parallel=1 (2 tests, 2 passed)
- flake8 django/utils/html.py tests/utils_tests/test_html.py tests/template_tests/filter_tests/test_urlize.py (no issues)

noa-lucent commented on django/utils/html.py:
```diff
@@ -283,8 +283,9 @@ def trim_punctuation(lead, middle, trail):
-                trail = middle[len(stripped):] + trail
-                middle = middle[:len(stripped) - len(middle_unescaped)]
+                trailing_count = len(middle_unescaped) - len(stripped)
+                trail = middle[-trailing_count:] + trail
```
[major] This still breaks encoded punctuation. Because we only slice `trailing_count` characters off the escaped token, a numeric entity that *is* the trailing punctuation (e.g. `&#33;` for '!') gets split apart. After this change, `urlize('Search for google.com/?q=&#33; and see.')` yields `<a href="http://google.com/?q=#3">google.com/?q=&#3</a>3;` — the `!` moved out of the URL, but the entity got truncated to `&#3` inside the link and the broken tail `3;` was appended after it. We need to move whole entities when they're the trimmed punctuation (walk backwards, consuming through the `&...;` block), otherwise we corrupt any escaped punctuation. Please fix the slicing logic and add coverage for this case.}

casey-brooks replied:
Thanks, Noa. I updated trim_punctuation to walk backward over semicolon-terminated entities that decode to punctuation so we move the whole entity at once, and added decimal/hex entity regressions on the utils and template paths. The new tests pass with the updated logic.

noa-lucent reviewed (changes requested):
Thanks for driving the regression coverage here. The new slicing logic still corrupts URLs when the trailing punctuation itself is represented as an HTML entity (e.g. `&#33;`). In that case we peel only part of the entity, yielding `http://google.com/?q=#3`. Please adjust the trimming to move whole entities and add a regression test for this path.

casey-brooks committed:
fix(utils.html): trim entity punctuation

Refs: #426

- consume semicolon-terminated entities decoding to trailing punctuation as single units
- extend utils/template urlize tests to cover decimal/hex entity punctuation cases

Addresses Noa's feedback on PR #431.

casey-brooks commented:
Updated Test & Lint Summary:
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py utils_tests.test_html.TestUtilsHtml.test_urlize --parallel=1 (1 test, 1 passed)
- PYTHONPATH=/workspace/django DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py template_tests.filter_tests.test_urlize --parallel=1 (41 tests, 41 passed)
- flake8 django/utils/html.py tests/utils_tests/test_html.py tests/template_tests/filter_tests/test_urlize.py (no issues)

rowan-stein reviewed (approved):
Approving after verifying the latest changes fix entity-trailing punctuation handling in urlize(). Tests for '&#33;' and '&#x21;' cases pass locally.

noa-lucent reviewed (approved):
The updated trimming logic now walks entire HTML entities that decode to trailing punctuation, so we no longer split numeric references. The new regression tests cover decimal/hex entities on both the utils and template code paths. Looks good.
